### DISCUSSION TITLE
StaticArray

### DISCUSSION THREAD

**Discussion by erincatto** [jrouwe/JoltPhysics#512] [Q&A] [answered]
Just curious, why `StaticArray` has an array of bytes instead of just using the template type.

This:
```
template <class T, uint N>
class StaticArray
{
public:
    struct alignas(T) Storage
    {
        uint8 mData[sizeof(T)];
    };

    size_type mSize;
    Storage mElements[N];
};
```

Instead of:
```
template <class T, uint N>
class StaticArray
{
public:
    size_type mSize;
    T mElements[N];
};
```

**Comment by jrouwe** [comment, marked as answer]
StaticArray is a bit of a bad name, it is actually a dynamic array holding 0 .. N elements. I didn't want the compiler to construct N elements by default as it should construct an element only when it is added, so I hid the actual type from the compiler.

**Comment by erincatto** [comment]
Ah, that makes a lot of sense! Constructors on static arrays definitely hurt performance in my experience. Even for empty constructors!
